### PR TITLE
Davidns/ch2141/implement frontend mocks for change learning object author

### PR DESCRIPTION
**Pull request by davidnso**
This story covers implementing the provided mocks client side for change author functionality. 

The user dropdown component from the builder was reused and modified to fit the needs of this new component. A completely separate dropdown component was created in the process.
Screen Shot 2020-06-11 at 5 30 48 PM 
Screen Shot 2020-06-11 at 5 30 59 PM
Screen Shot 2020-06-11 at 5 31 22 PM
Screen Shot 2020-06-11 at 5 31 33 PM

### CODE REVIEW

File: src/app/admin/components/change-author/change-author.component.html [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+       
+
+      </div>
+</div>
+<div class="content-header"> New Author </div>
+<div aria-live="assertive" *ngIf='selectAuthorFailure && selectAuthorFailure.length > 0' class="error" ><i  class="fas fa-info-circle"></i> {{ selectAuthorFailure || '' }} </div>
+    <clark-change-author-user-dropdown (newAuthor)="setSelectedAuthor($event)"></clark-change-author-user-dropdown>
+
+      <div class="btn-group center">
+        <button class="button good" (click)="toggleState(true)">Continue</button>
```
Comment (pzalep1) [inline src/app/admin/components/change-author/change-author.component.html:42]: Aria label please

File: src/app/admin/components/change-author/change-author.component.html [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+        <button class="button neutral" (activate)="close.emit()">Cancel</button>
```
Comment (pzalep1) [inline src/app/admin/components/change-author/change-author.component.html:44]: Aria label here as well.

File: src/app/admin/components/change-author/change-author.component.html [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+<div class="change-author-content">
+  <div class="content-header padding-extra">Are you sure you want to change this learning object's author?</div>
+  <div class="description sm padding-indent">Change the learning object author of "{{highlightedLearningObject.name | titlecase}}" from {{highlightedLearningObject.author.name | titlecase}} ({{highlightedLearningObject.author.organization | titlecase}}) to {{selectedAuthor._name | titlecase}} ({{selectedAuthor._organization | titlecase}})). </div>
+  <div class="description sm padding-indent">Please Check to ensure that the new author is the correct user that you would like to change authorship of. Note that 
+    changing the author will add the previous author onto the list of contributors for the object. 
+  </div>
+
+</div>
+<div class="change-author-confirmation">
+  <input (change)="toggleConsent($event)" name='change-author-checkbox' id='change-author-checkbox' type="checkbox">
```
Comment (pzalep1) [inline src/app/admin/components/change-author/change-author.component.html:64]: Aria labeled by

File: src/app/admin/components/change-author/change-author.component.html [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  <label for='change-author-checkbox' class="description xs">I have checked that the user I want to change authorship to is the correct user</label>
+</div>
+<div class="btn-group center">
+
+  <button class="button good">Change Author</button>
```
Comment (pzalep1) [inline src/app/admin/components/change-author/change-author.component.html:69]: Aria label on both of these

File: src/app/admin/components/change-author/change-author.component.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  styleUrls: ['./change-author.component.scss']
+})
+export class ChangeAuthorComponent implements OnInit {
+
+  finalStage = false;
+  selectAuthorFailure: string;
+  consentGiven = false;
+  failureTimer;
+  constructor() { }
+  @Input() highlightedLearningObject: LearningObject;
```
Comment (pzalep1) [inline src/app/admin/components/change-author/change-author.component.ts:16]: These should go above the constructor with the other variable declarations.

File: src/app/admin/components/change-author/components/change-author-user-dropdown/change-author-user-dropdown.component.html [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    <div *ngIf="selectedAuthor && selectedAuthor.username" class="author-details-container">
+      <div class="left">
+        <span class="user-dropdown__navbar-gravatar image" [ngStyle]="{'background-image': 'url(' + userService.getGravatarImage(selectedAuthor.email, 200) + ')'}"></span>
+        <span class="details">
+              <p>{{selectedAuthor.username}}</p>
+              <p>{{selectedAuthor.organization}} - {{selectedAuthor.email}}</p>
+        </span>
+      </div>
+      <div (click)="toggleAuthor()" class="right"><i class="fas fa-times"></i></div>
```
Comment (pzalep1) [inline src/app/admin/components/change-author/components/change-author-user-dropdown/change-author-user-dropdown.component.html:18]: Should be a button

File: src/app/admin/components/change-author/components/change-author-user-dropdown/change-author-user-dropdown.component.scss [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    font-size: $small;
+    color: $medium-grey
+}
+
+.user-dropdown {
+    position: relative;
+    min-height: 150px;
+  }
+  .input-container{
+    background-color: #F1F1F1;
```
Comment (pzalep1) [inline src/app/admin/components/change-author/components/change-author-user-dropdown/change-author-user-dropdown.component.scss:42]: There is a placeholder value in the vars file

File: src/app/admin/components/change-author/change-author.component.scss [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    &:after {
+      content: '';
+      display: block;
+      padding-top: 100%;
+    }
+  }
+
+  .change-author-confirmation { 
+    display: flex;
+    border: 1px solid #d1d1d1;
```
Comment (pzalep1) [inline src/app/admin/components/change-author/change-author.component.scss:148]: Is there is one of the greys that would work here? $light-grey might be close.

File: src/app/admin/components/change-author/change-author.component.scss [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    padding: 10px 0px;
+
+  }
+
+  .padding-indent{ 
+    padding: 10px 10px
+  }
+
+  .description{ 
+    color: #424242;
```
Comment (pzalep1) [inline src/app/admin/components/change-author/change-author.component.scss:170]: Same thing here with the grey, $dark-grey might work here

File: src/app/admin/components/change-author/components/change-author-user-dropdown/change-author-user-dropdown.component.ts [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+import { AuthService } from 'app/core/auth.service';
+import { Subject } from 'rxjs';
+import { debounceTime, takeUntil } from 'rxjs/operators';
+ 
+@Component({
+  selector: 'clark-change-author-user-dropdown',
+  templateUrl: './change-author-user-dropdown.component.html',
+  styleUrls: ['./change-author-user-dropdown.component.scss'],
+})
+export class ChangeAuthorUserDropdownComponent implements OnInit {
```
Comment (pzalep1) [inline src/app/admin/components/change-author/components/change-author-user-dropdown/change-author-user-dropdown.component.ts:20]: This should implement onDestroy to take care of the observables

File: src/app/admin/components/learning-object-list-item/learning-object-list-item.component.ts [hunk 11]
```diff
@@ -68,6 +71,10 @@ export class LearningObjectListItemComponent implements OnChanges {
     this.meatballOpen = value;
   }
 
+  toggleChangeAuthorModal(value: boolean){
+    this.showChangeAuthor = value;
+    console.log(this.learningObject);
```
Comment (pzalep1) [inline src/app/admin/components/learning-object-list-item/learning-object-list-item.component.ts:76]: :fire: Console log

File: src/app/shared/components/shared-components.module.ts [hunk 12]
```diff
@@ -53,6 +54,7 @@ import { SkipLinkComponent } from './skip-link/skip-link.component';
     ToggleSwitchComponent,
     UserCardComponent,
     SkipLinkComponent,
+
```
Comment (pzalep1) [inline src/app/shared/components/shared-components.module.ts:57]: Why is there a new line here?

File: src/app/shared/components/shared-components.module.ts [hunk 13]
```diff
@@ -26,6 +26,7 @@ import { UserCardComponent } from './user-card/user-card.component';
 import { SkipLinkComponent } from './skip-link/skip-link.component';
 
 
+
```
Comment (pzalep1) [inline src/app/shared/components/shared-components.module.ts:29]: Why is there a new line?

File: src/app/admin/components/change-author/components/change-author-user-dropdown/change-author-user-dropdown.component.ts [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    // subscribe to the search input and fire search after debounce
+    this.userSearchInput$
+      .pipe(debounceTime(650), takeUntil(this.destroyed$))
+      .subscribe((val: string) => {
+        this.findUser(val.trim());
+      });
+
+    // TODO is there a better way to do this?
+    // I want two events to occur when this event fires. 1) after a debounce, perform search
+    // 2) immediately (without debounce) show the popup
```
Comment (pzalep1) [inline src/app/admin/components/change-author/components/change-author-user-dropdown/change-author-user-dropdown.component.ts:61]: Can I get a little more context on this? We only want to show the drop down if there are users to show correct?

**Comment by Olaniyan99** [comment]
this looks good, no major comments just the changes Paige requested for. looks real good, i like!

File: src/app/admin/components/change-author/components/change-author-user-dropdown/change-author-user-dropdown.component.ts (hunk 14 above)
Comment (davidnso) [reply src/app/admin/components/change-author/components/change-author-user-dropdown/change-author-user-dropdown.component.ts:61]: This comment was in the original user dropdown component, So Nick will likely be able to give more context on it.

Comment (pzalep1) [reply src/app/admin/components/change-author/components/change-author-user-dropdown/change-author-user-dropdown.component.ts:61]: Gotcha. Thanks

**Pull request merged**